From d33216fa76d9db0b0f4c3c7863ddd0cdbbb0ea84 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Fri, 31 Oct 2025 15:09:33 +0100
Subject: [PATCH] usb: dwc2: avoid wakeup interrupt race while suspending

It may happen a remote wakeup interrupt is generated, during system
wide PM suspend. In case the gadget is connected (hence the remote
wakeup may fire), abort the suspend sequence, by calling
pm_wakeup_hard_event().

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: Ieac59af94498a77973ac70ac0e3c292f1eaef48a
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/512986
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
---
 drivers/usb/dwc2/core_intr.c | 4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/usb/dwc2/core_intr.c
+++ b/drivers/usb/dwc2/core_intr.c
@@ -419,6 +419,10 @@ static void dwc2_handle_wakeup_detected_
 	}
 
 	if (dwc2_is_device_mode(hsotg)) {
+		/* Prevents remote wakeup interrupt race while suspending */
+		if (dwc2_is_device_connected(hsotg))
+			pm_wakeup_hard_event(hsotg->dev);
+
 		dev_dbg(hsotg->dev, "DSTS=0x%0x\n",
 			dwc2_readl(hsotg, DSTS));
 		if (hsotg->lx_state == DWC2_L2) {
